### PR TITLE
ngcc bundle PR including implementation work for other formats & integration test

### PR DESCRIPTION
**Pull request by petebacondarwin**
This sits on top of #25203

### CODE REVIEW

File: packages/compiler-cli/src/ngtsc/metadata/src/resolver.ts [hunk 1]
```diff
@@ -17,7 +17,7 @@ import * as ts from 'typescript';
 
 import {ClassMemberKind, ReflectionHost} from '../../host';
 
-const TS_DTS_EXTENSION = /(\.d)?\.ts$/;
+const TS_DTS_JS_EXTENSION = /(\.d)?\.(ts|js)$/;
```
Comment (JoostK) [inline packages/compiler-cli/src/ngtsc/metadata/src/resolver.ts]: @alxhub For a file ending in `.d.js`, all of `.d.ts` will be removed below, whereas I'd expect `.d` to be kept.

Comment (petebacondarwin) [reply packages/compiler-cli/src/ngtsc/metadata/src/resolver.ts]: Good catch @JoostK .

I think we should go with `/((\.d)?\.ts|js)$/`

Comment (JoostK) [reply packages/compiler-cli/src/ngtsc/metadata/src/resolver.ts]: @petebacondarwin actually: `/((\.d)?\.ts|\.js)$/`

Comment (petebacondarwin) [reply packages/compiler-cli/src/ngtsc/metadata/src/resolver.ts]: Oops. Yes :-)

**Comment by alxhub** [comment]
http://test/OCL:208201706:BASE:208201747:1533908396120:7553c6e7

**Review by mhevery — APPROVED** [review]

File: packages/compiler-cli/src/ngcc/src/main.ts [hunk 2]
```diff
@@ -5,17 +5,84 @@
...
-  // TODO: error/warning logging/handling etc
+  const formats = args[0] ? args[0].split(',') : ['fesm2015', 'esm2015', 'fesm5', 'esm5'];
+  const packagePaths = args[1] ? [resolve(args[1])] : findPackagesToCompile();
+  const targetPath = args[2] ? args[2] : 'node_modules';
 
   const transformer = new PackageTransformer();
-  transformer.transform(packagePath, 'fesm2015');
+  packagePaths.forEach(packagePath => {
+    formats.forEach(format => {
+      console.warn(`Compiling ${packagePath} : ${format}`);
```
Comment (gkalpak) [inline packages/compiler-cli/src/ngcc/src/main.ts]: If this is temporary, add a TODO comment to remove it. If not, then it should use `console.log` instead.

File: packages/compiler-cli/src/ngcc/src/transform/package_transformer.ts [hunk 3]
```diff
@@ -5,77 +5,105 @@
...
+import {Fesm2015ReflectionHost} from '../host/fesm2015_host';
 import {NgccReflectionHost} from '../host/ngcc_host';
 import {Esm2015FileParser} from '../parsing/esm2015_parser';
 import {Esm5FileParser} from '../parsing/esm5_parser';
 import {FileParser} from '../parsing/file_parser';
-import {getEntryPoints} from '../parsing/utils';
 import {Esm2015Renderer} from '../rendering/esm2015_renderer';
 import {Esm5Renderer} from '../rendering/esm5_renderer';
 import {FileInfo, Renderer} from '../rendering/renderer';
+import {isDefined} from '../utils';
```
Comment (gkalpak) [inline packages/compiler-cli/src/ngcc/src/transform/package_transformer.ts]: Unused import.

**Review by gkalpak — APPROVED** [review]
Some minor comments. LGTM otherwise.

**Comment by gkalpak** [comment]
Rebased on top of #25203.

File: packages/compiler-cli/src/ngcc/src/main.ts [hunk 4]
```diff
@@ -5,23 +5,84 @@
...
+
+/**
+ * Look for packages that need to be compiled.
+ * The function will recurse into folders that start with `@...`, e.g. `@angular/...`.
+ *  Without an argument it starts at `node_modules`.
+ */
+function findPackagesToCompile(folder: string = 'node_modules'): string[] {
+  const path = resolve(folder);
+  const packagesToCompile: string[] = [];
+  readdirSync(path).filter(p => !p.startsWith('.')).forEach(p => {
```
[the remark this answers is not in the document]
Comment (alxhub) [reply packages/compiler-cli/src/ngcc/src/main.ts]: Good call - done.

File: packages/compiler-cli/src/ngcc/src/main.ts [hunk 5]
```diff
@@ -5,23 +5,84 @@
...
+  // TODO: avoid if packageJson contains built marker
+  const metadataPath =
+      resolve(folderPath, packageJson.typings.replace(/\.d\.ts$/, '.metadata.json'));
+  return existsSync(metadataPath);
+}
+
+/**
+ * Look for packages that need to be compiled.
+ * The function will recurse into folders that start with `@...`, e.g. `@angular/...`.
+ *  Without an argument it starts at `node_modules`.
```
[the remark this answers is not in the document]
Comment (alxhub) [reply packages/compiler-cli/src/ngcc/src/main.ts]: Fixed.

File: packages/compiler-cli/src/ngcc/src/main.ts [hunk 6]
```diff
@@ -5,23 +5,84 @@
...
+ * - do not contain a package.json
+ * - do not have a typings property in package.json
+ * - do not have an appropriate metadata.json file
+ *
+ * @param folderPath The absolute path to the folder.
+ * @param folderName The final segment of the folder path.
+ */
+function hasMetadataFile(folderPath: string): boolean {
+  const folderName = basename(folderPath);
+  if (lstatSync(folderPath).isSymbolicLink() || folderName === 'node_modules') {
```
[the remark this answers is not in the document]
Comment (alxhub) [reply packages/compiler-cli/src/ngcc/src/main.ts]: Done.

File: packages/compiler-cli/src/ngcc/src/main.ts [hunk 7]
```diff
@@ -5,23 +5,84 @@
...
+ * We do not care about folders that are:
+ *
+ * - symlinks
+ * - node_modules
+ * - do not contain a package.json
+ * - do not have a typings property in package.json
+ * - do not have an appropriate metadata.json file
+ *
+ * @param folderPath The absolute path to the folder.
+ * @param folderName The final segment of the folder path.
```
[the remark this answers is not in the document]
Comment (alxhub) [reply packages/compiler-cli/src/ngcc/src/main.ts]: Removed.

File: packages/compiler-cli/src/ngcc/src/main.ts [hunk 8]
```diff
@@ -5,23 +5,84 @@
...
+/**
+ * Look for packages that need to be compiled.
+ * The function will recurse into folders that start with `@...`, e.g. `@angular/...`.
+ *  Without an argument it starts at `node_modules`.
+ */
+function findPackagesToCompile(folder: string = 'node_modules'): string[] {
+  const path = resolve(folder);
+  const packagesToCompile: string[] = [];
+  readdirSync(path).filter(p => !p.startsWith('.')).forEach(p => {
+    const packagePath = resolve(path, p);
```
[the remark this answers is not in the document]
Comment (alxhub) [reply packages/compiler-cli/src/ngcc/src/main.ts]: Done, and converted to `path.posix` for extra correctness :)

File: packages/compiler-cli/src/ngcc/src/main.ts [hunk 9]
```diff
@@ -5,23 +5,84 @@
...
+      resolve(folderPath, packageJson.typings.replace(/\.d\.ts$/, '.metadata.json'));
+  return existsSync(metadataPath);
+}
+
+/**
+ * Look for packages that need to be compiled.
+ * The function will recurse into folders that start with `@...`, e.g. `@angular/...`.
+ *  Without an argument it starts at `node_modules`.
+ */
+function findPackagesToCompile(folder: string = 'node_modules'): string[] {
```
[the remark this answers is not in the document]
Comment (alxhub) [reply packages/compiler-cli/src/ngcc/src/main.ts]: It's definitely not what we want long term, but it'll do for now.

File: packages/compiler-cli/src/ngcc/src/transform/utils.ts [hunk 10]
```diff
@@ -21,44 +21,19 @@ import {isDefined} from '../utils';
  */
```
[the remark this answers is not in the document]
Comment (alxhub) [reply packages/compiler-cli/src/ngcc/src/transform/utils.ts]: How has the implementation changed? This is a new file.

File: packages/compiler-cli/src/ngcc/src/host/esm2015_host.ts [hunk 11]
```diff
@@ -6,420 +6,37 @@
...
-      if (symbol.exports && symbol.exports.has(DECORATORS)) {
-        // Symbol of the identifier for `SomeDirective.decorators`.
-        const decoratorsSymbol = symbol.exports.get(DECORATORS) !;
-        const decoratorsIdentifier = decoratorsSymbol.valueDeclaration;
+  getGenericArityOfClass(clazz: ts.Declaration): number|null {
+    if (ts.isClassDeclaration(clazz) && clazz.name) {
+      const sourcePath = clazz.getSourceFile();
+      const dtsPath = this.dtsMapper.getDtsFileNameFor(sourcePath.fileName);
+      const dtsContents = readFileSync(dtsPath, 'utf8');
+      const dtsFile = ts.createSourceFile(
```
[the remark this answers is not in the document]
Comment (alxhub) [reply packages/compiler-cli/src/ngcc/src/host/esm2015_host.ts]: Done.

File: packages/compiler-cli/src/ngcc/test/host/esm2015_host_spec.ts [hunk 12]
```diff
@@ -6,1109 +6,50 @@
...
-      expect(host.isClass(node)).toBe(true);
-    });
-
-    it('should return false if a given node is a TS function declaration', () => {
-      const program = makeProgram(FOO_FUNCTION_FILE);
-      const host = new Esm2015ReflectionHost(program.getTypeChecker());
-      const node = getDeclaration(program, FOO_FUNCTION_FILE.name, 'foo', ts.isFunctionDeclaration);
-      expect(host.isClass(node)).toBe(false);
+  describe('getGenericArityOfClass()', () => {
+    it('should return null', () => {
```
[the remark this answers is not in the document]
Comment (alxhub) [reply packages/compiler-cli/src/ngcc/test/host/esm2015_host_spec.ts]: Named properly.

File: packages/compiler-cli/src/ngcc/src/main.ts (hunk 2 above)
Comment (alxhub) [reply packages/compiler-cli/src/ngcc/src/main.ts]: Done.

File: packages/compiler-cli/src/ngcc/src/transform/package_transformer.ts (hunk 3 above)
Comment (alxhub) [reply packages/compiler-cli/src/ngcc/src/transform/package_transformer.ts]: Removed.

File: packages/compiler-cli/src/ngcc/src/transform/utils.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (gkalpak) [reply packages/compiler-cli/src/ngcc/src/transform/utils.ts]: This file is basically `src/parsing/utils.ts` renamed to `src/transform/utils.ts` plus some changes. The changes are significant enough that git does not detect this as a rename.

**Comment by petebacondarwin** [comment]
Regarding https://github.com/angular/angular/pull/25406#issuecomment-414691378, I have split the commit into two:

* refactor(ivy): rename `Esm2015ReflectionHost` to `Fesm2015ReflectionH… 098e066
* feat(ivy): implement `Esm5ReflectionHost.getGenericArityOfClass()` 52a19b9

**Review by alxhub — APPROVED** [review]

**Pull request closed without merging**